File: library_zap/src/main/java/com/dolin/zap/lifecycle/ZapLifecycle.java
package com.dolin.zap.lifecycle;

import android.app.Activity;
import android.app.Application;
import android.os.Bundle;

import com.dolin.comm.impl.R4LogHandler;

/**
 * @author #Suyghur.
 * Created on 2021/05/19
 */
public class ZapLifecycle {

    private static volatile ZapLifecycle mInstance = null;

    private Application application = null;
    private R4LogHandler r4LogHandler = null;


    private final Application.ActivityLifecycleCallbacks callback = new Application.ActivityLifecycleCallbacks() {
        @Override
        public void onActivityCreated(Activity activity, Bundle savedInstanceState) {

        }

        @Override
        public void onActivityStarted(Activity activity) {

        }

        @Override
        public void onActivityResumed(Activity activity) {

        }

        @Override
        public void onActivityPaused(Activity activity) {
            if (r4LogHandler != null) {
                r4LogHandler.asyncFlush();
            }
        }

        @Override
        public void onActivityStopped(Activity activity) {

        }

        @Override
        public void onActivitySaveInstanceState(Activity activity, Bundle outState) {

        }

        @Override
        public void onActivityDestroyed(Activity activity) {

        }
    };


    private ZapLifecycle() {

    }

    public static ZapLifecycle getInstance() {
        if (mInstance == null) {
            synchronized (ZapLifecycle.class) {
                if (mInstance == null) {
                    mInstance = new ZapLifecycle();
                }
            }
        }
        return mInstance;
    }

    public void registerZapLifeCallback(Application application, R4LogHandler r4LogHandler) {
        this.application = application;
        this.r4LogHandler = r4LogHandler;
        application.registerActivityLifecycleCallbacks(callback);
    }

    public void unregisterZapLifeCallback() {
        if (application != null) {
            application.unregisterActivityLifecycleCallbacks(callback);
            this.application = null;
        }
        if (r4LogHandler != null) {
            this.r4LogHandler = null;
        }
    }
}

File: library_zap/src/main/java/com/suyghur/dolin/zap/lifecycle/ZapLifecycle.kt
package com.suyghur.dolin.zap.lifecycle

import android.app.Activity
import android.app.Application
import android.os.Bundle
import com.suyghur.dolin.zap.impl.Record2MMap


/**
 * @author #Suyghur.
 * Created on 4/8/21
 */
object ZapLifecycle {

    private var record2MMap: Record2MMap? = null
    private var application: Application? = null

    private val callback = object : Application.ActivityLifecycleCallbacks {
        override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {
        }

        override fun onActivityStarted(activity: Activity) {
        }

        override fun onActivityResumed(activity: Activity) {
        }

        override fun onActivityPaused(activity: Activity) {
            record2MMap?.asyncFlush()
        }

        override fun onActivityStopped(activity: Activity) {
        }

        override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {
        }

        override fun onActivityDestroyed(activity: Activity) {
        }
    }

    fun registerZapLifeCallback(application: Application, record2MMap: Record2MMap) {
        ZapLifecycle.application = application
        ZapLifecycle.record2MMap = record2MMap
        application.registerActivityLifecycleCallbacks(callback)
    }

    fun unregisterZapLifeCallback() {
        application?.apply {
            unregisterActivityLifecycleCallbacks(callback)
            application = null
        }
        record2MMap?.apply {
            record2MMap = null
        }

    }
}